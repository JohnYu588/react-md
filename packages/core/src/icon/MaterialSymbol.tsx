import { forwardRef, type AriaAttributes, type HTMLAttributes } from "react";
import { type MaterialSymbolName } from "./material.js";
import {
  getFontVariationSettings,
  type MaterialSymbolCustomization,
  type MaterialSymbolFill,
  type MaterialSymbolGrade,
  type MaterialSymbolOpticalSize,
  type MaterialSymbolWeight,
} from "./materialConfig.js";
import { icon, type MaterialSymbolClassNameOptions } from "./styles.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-symbol-fill"?: MaterialSymbolFill;
    "--rmd-symbol-wght"?: MaterialSymbolWeight;
    "--rmd-symbol-grad"?: MaterialSymbolGrade;
    "--rmd-symbol-opsz"?: MaterialSymbolOpticalSize;
  }
}

/**
 * @since 6.0.0
 */
export interface MaterialSymbolProps
  extends HTMLAttributes<HTMLSpanElement>,
    MaterialSymbolCustomization,
    Partial<MaterialSymbolClassNameOptions> {
  /** @defaultValue `true` */
  "aria-hidden"?: AriaAttributes["aria-hidden"];
  name: MaterialSymbolName;
  children?: never;
}

/**
 * This is a convenience component that provides autocomplete for all the
 * available material symbols via the `name` prop.
 *
 * Note: You might notice IDE slowdowns for files that use this component since
 * there are so many icons available. If it becomes an issue, just stop using
 * this component and define the icons inline instead.
 *
 * @example Simple Example
 * ```tsx
 * import { MaterialSymbol } from "@react-md/core";
 * import type { ReactElement } from "react";
 * import { createRoot } from "react-dom/client";
 *
 * function Example(): ReactElement {
 *   return (
 *     <>
 *       <MaterialSymbol symbol="close" />
 *       <MaterialSymbol symbol="tune" type="outline" />
 *
 *       <MaterialSymbol symbol="add" type="round" />
 *       <MaterialSymbol symbol="air" type="sharp" />
 *     </>
 *   );
 * }
 *
 * const root = createRoot(document.getElementById("root"));
 * root.render(<App />);
 * ```
 *
 * @since 6.0.0
 */
export const MaterialSymbol = forwardRef<HTMLSpanElement, MaterialSymbolProps>(
  function MaterialSymbol(props, ref) {
    const {
      "aria-hidden": ariaHidden = true,
      className,
      name: symbol,
      style: propStyle,
      family: propFamily,
      fill,
      weight,
      grade,
      opticalSize,
      theme,
      dense,
      ...remaining
    } = props;
    const { style, family } = getFontVariationSettings({
      style: propStyle,
      fill,
      weight,
      grade,
      opticalSize,
      family: propFamily,
    });

    return (
      <span
        {...remaining}
        aria-hidden={ariaHidden}
        ref={ref}
        style={style}
        className={icon({
          type: "symbol",
          family,
          theme,
          dense,
          className,
        })}
      >
        {symbol}
      </span>
    );
  }
);
