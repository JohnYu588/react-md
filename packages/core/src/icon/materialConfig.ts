import { type CSSProperties } from "react";

import {
  type MaterialIconFamily,
  type MaterialSymbolFamily,
} from "./material.js";

/**
 * @since 6.0.0
 */
export interface MaterialIconCustomization {
  /**
   * @defaultValue `"filled"`
   */
  iconFamily?: MaterialIconFamily;
}

/**
 * @since 6.0.0
 */
export type MaterialIconConfiguration = Required<MaterialIconCustomization>;

/**
 * Fill gives you the ability to modify the default icon style. A single icon
 * can render both unfilled and filled states.
 *
 * To convey a state transition, use the fill axis for animation or
 * interaction. The values are 0 for default or 1 for completely filled. Along
 * with the weight axis, the fill also impacts the look of the icon.
 *
 * @since 6.0.0
 */
export type MaterialSymbolFill = 0 | 1;

/**
 * Weight defines the symbol’s stroke weight, with a range of weights between
 * thin (100) and bold (700). Weight can also affect the overall size of the
 * symbol.
 *
 * @since 6.0.0
 */
export type MaterialSymbolWeight = 100 | 200 | 300 | 400 | 500 | 600 | 700;

/**
 * Weight and grade affect a symbol’s thickness. Adjustments to grade are more
 * granular than adjustments to weight and have a small impact on the size of
 * the symbol.
 *
 * Grade is also available in some text fonts. You can match grade levels
 * between text and symbols for a harmonious visual effect. For example, if
 * the text font has a -25 grade value, the symbols can match it with a
 * suitable value, say -25.
 *
 * You can use grade for different needs:
 *
 * - Low emphasis (e.g. -25 grade): To reduce glare for a light symbol on a
 *   dark background, use a low grade.
 * - High emphasis (e.g. 200 grade): To highlight a symbol, increase the
 *   positive grade.
 *
 * @since 6.0.0
 */
export type MaterialSymbolGrade = -25 | 0 | 200;

/**
 * Optical Sizes range from 20dp to 48dp.
 *
 * For the image to look the same at different sizes, the stroke weight
 * (thickness) changes as the icon size scales. Optical Size offers a way to
 * automatically adjust the stroke weight when you increase or decrease the
 * symbol size.
 *
 * @since 6.0.0
 */
export type MaterialSymbolOpticalSize = 20 | 24 | 40 | 48;

/**
 * The comments for each customizable part was copied from
 * https://fonts.google.com/icons?icon.set=Material+Symbols and clicking the
 * info icon next to each property.
 *
 * @see https://fonts.google.com/icons?icon.set=Material+Symbols
 * @since 6.0.0
 */
export interface MaterialSymbolCustomization {
  /**
   * Fill gives you the ability to modify the default icon style. A single icon
   * can render both unfilled and filled states.
   *
   * To convey a state transition, use the fill axis for animation or
   * interaction. The values are 0 for default or 1 for completely filled. Along
   * with the weight axis, the fill also impacts the look of the icon.
   *
   * @defaultValue `0`
   */
  fill?: MaterialSymbolFill;

  /**
   * Weight defines the symbol’s stroke weight, with a range of weights between
   * thin (100) and bold (700). Weight can also affect the overall size of the
   * symbol.
   *
   * @defaultValue `400`
   */
  weight?: MaterialSymbolWeight;

  /**
   * Weight and grade affect a symbol’s thickness. Adjustments to grade are more
   * granular than adjustments to weight and have a small impact on the size of
   * the symbol.
   *
   * Grade is also available in some text fonts. You can match grade levels
   * between text and symbols for a harmonious visual effect. For example, if
   * the text font has a -25 grade value, the symbols can match it with a
   * suitable value, say -25.
   *
   * You can use grade for different needs:
   *
   * - Low emphasis (e.g. -25 grade): To reduce glare for a light symbol on a
   *   dark background, use a low grade.
   * - High emphasis (e.g. 200 grade): To highlight a symbol, increase the
   *   positive grade.
   *
   * @defaultValue `0`
   */
  grade?: MaterialSymbolGrade;

  /**
   * Optical Sizes range from 20dp to 48dp.
   *
   * For the image to look the same at different sizes, the stroke weight
   * (thickness) changes as the icon size scales. Optical Size offers a way to
   * automatically adjust the stroke weight when you increase or decrease the
   * symbol size.
   *
   * @defaultValue `48`
   */
  opticalSize?: MaterialSymbolOpticalSize;

  /**
   * @defaultValue `"outlined"`
   */
  family?: MaterialSymbolFamily;
}

/**
 * @since 6.0.0
 */
export type MaterialSymbolConfiguration = Required<MaterialSymbolCustomization>;

/**
 * @since 6.0.0
 */
export interface MaterialConfiguration
  extends MaterialIconConfiguration,
    MaterialSymbolConfiguration {}

/**
 * @since 6.0.0
 */
export const MATERIAL_CONFIG: MaterialConfiguration = {
  fill: 0,
  weight: 400,
  grade: 0,
  opticalSize: 48,
  iconFamily: "filled",
  family: "outlined",
};

/**
 * @since 6.0.0
 */
export interface MaterialSymbolFontVariationSettings {
  style?: CSSProperties;
  family: MaterialSymbolFamily;
}

interface ApplyOptions {
  name: "fill" | "wght" | "grad" | "opsz";
  value: number | undefined;
  defaultValue: number;
  style: CSSProperties | undefined;
}

const applyVar = (options: ApplyOptions): CSSProperties | undefined => {
  const { name, value, defaultValue, style } = options;
  if (typeof value !== "undefined" && value !== defaultValue) {
    const varName = `--rmd-symbol-${name}` as const;
    return {
      ...style,
      [varName]: value,
    };
  }

  return style;
};

/**
 * @since 6.0.0
 * @internal
 */
export function getFontVariationSettings(
  options: MaterialSymbolCustomization & { style?: CSSProperties }
): MaterialSymbolFontVariationSettings {
  const {
    family = MATERIAL_CONFIG.family,
    fill,
    grade,
    opticalSize,
    weight,
  } = options;
  let { style } = options;
  style = applyVar({
    style,
    name: "fill",
    value: fill,
    defaultValue: MATERIAL_CONFIG.fill,
  });
  style = applyVar({
    style,
    name: "grad",
    value: grade,
    defaultValue: MATERIAL_CONFIG.grade,
  });
  style = applyVar({
    style,
    name: "opsz",
    value: opticalSize,
    defaultValue: MATERIAL_CONFIG.opticalSize,
  });
  style = applyVar({
    style,
    name: "wght",
    value: weight,
    defaultValue: MATERIAL_CONFIG.weight,
  });

  return { style, family };
}
