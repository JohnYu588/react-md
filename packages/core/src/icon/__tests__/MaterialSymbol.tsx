import { beforeEach, describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { render, screen } from "../../test-utils/index.js";
import { MaterialSymbol } from "../MaterialSymbol.js";
import { MATERIAL_CONFIG } from "../materialConfig.js";

describe("MaterialSymbol", () => {
  beforeEach(() => {
    MATERIAL_CONFIG.fill = 0;
    MATERIAL_CONFIG.weight = 400;
    MATERIAL_CONFIG.grade = 0;
    MATERIAL_CONFIG.opticalSize = 24;
    MATERIAL_CONFIG.family = "outlined";
  });

  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLSpanElement>();
    const props = {
      "data-testid": "icon",
      ref,
      name: "air",
    } as const;

    const { rerender } = render(<MaterialSymbol {...props} />);

    const icon = screen.getByTestId("icon");
    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
    expect(ref.current).toBe(icon);
    expect(icon).toMatchSnapshot();

    rerender(
      <MaterialSymbol
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(icon).toMatchSnapshot();
  });

  it("should support overriding the symbol configuration through the MATERIAL_CONFIG object", () => {
    const props = {
      "data-testid": "icon",
      name: "air",
    } as const;
    const { rerender } = render(<MaterialSymbol {...props} />);
    const icon = screen.getByTestId("icon");
    expect(icon).toHaveClass("material-symbols-outlined");
    expect(icon).toMatchSnapshot();

    MATERIAL_CONFIG.fill = 1;
    MATERIAL_CONFIG.grade = -25;
    MATERIAL_CONFIG.family = "rounded";
    MATERIAL_CONFIG.opticalSize = 48;
    MATERIAL_CONFIG.weight = 100;

    rerender(<MaterialSymbol {...props} />);
    expect(icon).toHaveClass("material-symbols-rounded");
    expect(icon).toMatchSnapshot();
  });

  it("should use inline styles for font variation settings when they do not match the MATERIAL_CONFIG", () => {
    const props = {
      "data-testid": "icon",
      name: "air",
    } as const;
    const { rerender } = render(<MaterialSymbol {...props} />);
    const icon = screen.getByTestId("icon");
    expect(icon).toHaveClass("material-symbols-outlined");
    expect(icon.style.fontVariationSettings).toBe("");

    rerender(<MaterialSymbol {...props} fill={1} />);
    expect(icon.style.fontVariationSettings).toBe(
      '"FILL" 1, "wght" 400, "GRAD" 0, "opsz" 24'
    );

    rerender(
      <MaterialSymbol {...props} fill={1} grade={-25} opticalSize={48} />
    );
    expect(icon.style.fontVariationSettings).toBe(
      '"FILL" 1, "wght" 400, "GRAD" -25, "opsz" 48'
    );
  });

  it("should use the prop style fontVariationSettings if it exists", () => {
    const props = {
      "data-testid": "icon",
      name: "air",
      style: {
        fontVariationSettings: '"FILL" 1, "wght" 400, "GRAD" -25, "opsz" 48',
      },
    } as const;
    render(<MaterialSymbol {...props} />);
    const icon = screen.getByTestId("icon");

    expect(icon.style.fontVariationSettings).toBe(
      '"FILL" 1, "wght" 400, "GRAD" -25, "opsz" 48'
    );
  });
});
