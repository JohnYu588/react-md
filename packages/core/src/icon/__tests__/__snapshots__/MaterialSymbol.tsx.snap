// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialSymbol should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-outlined"
  data-testid="icon"
>
  air
</span>
`;

exports[`MaterialSymbol should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-outlined custom-class-name"
  data-testid="icon"
  style="color: white;"
>
  air
</span>
`;

exports[`MaterialSymbol should support overriding the symbol configuration through the MATERIAL_CONFIG object 1`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-outlined"
  data-testid="icon"
>
  air
</span>
`;

exports[`MaterialSymbol should support overriding the symbol configuration through the MATERIAL_CONFIG object 2`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-rounded"
  data-testid="icon"
>
  air
</span>
`;

exports[`MaterialSymbol should use inline styles for font variation settings when they do not match the MATERIAL_CONFIG 1`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-outlined"
  data-testid="icon"
  style="--rmd-symbol-fill: 1;"
>
  air
</span>
`;

exports[`MaterialSymbol should use inline styles for font variation settings when they do not match the MATERIAL_CONFIG 2`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--symbol material-symbols-outlined"
  data-testid="icon"
  style="--rmd-symbol-fill: 1; --rmd-symbol-grad: -25; --rmd-symbol-opsz: 48;"
>
  air
</span>
`;
